src/Firebase.js

```diff
@@ -0,0 +1,158 @@
+  pickBy(getQuery(props, ref, app), prop => typeof prop === 'function')
+
+export default class Firebase extends React.Component {
+  static propTypes = {
+    render: PropTypes.func.isRequired,
+    query: PropTypes.oneOfType([PropTypes.object, PropTypes.string, PropTypes.func]),
+    children: PropTypes.oneOfType([PropTypes.object, PropTypes.string, PropTypes.func]),
+    firebaseApp: PropTypes.instanceOf(firebase.app.App), // eslint-disable-line react/no-unused-prop-types
+  }
+
+  static contextTypes = {
+    firebaseApp: PropTypes.instanceOf(firebase.app.App),
+  }
+
+  state = {
+    subscriptionsState: {},
+    connected: false,
+  }
+
+  componentDidMount() {
+    this.mounted = true
+    const subscriptions = getSubscriptions(this.props, this.getRef(), this.getFirebaseApp())
+    this.subscribe(subscriptions)
+  }
+
+  componentWillReceiveProps(nextProps, nextContext) {
+    const app = this.getFirebaseApp(nextProps, nextContext)
+    const ref = this.getRef(nextProps, nextContext)
+    const subscriptions = getSubscriptions(this.props, ref, app)
+    const nextSubscriptions = getSubscriptions(nextProps, ref, app)
+    const addedSubscriptions = pickBy(nextSubscriptions, (path, key) => !subscriptions[key])
+    const removedSubscriptions = pickBy(subscriptions, (path, key) => !nextSubscriptions[key])
+    const changedSubscriptions = pickBy(
+      nextSubscriptions,
+      (path, key) => subscriptions[key] && subscriptions[key] !== path,
+    )
+
+    this.unsubscribe({ ...removedSubscriptions, ...changedSubscriptions })
+    this.subscribe({ ...addedSubscriptions, ...changedSubscriptions }, ref)
+  }
+
+  componentWillUnmount() {
+    this.mounted = false
+
+    if (this.listeners) {
+      this.unsubscribe(this.listeners)
+    }
+  }
+
+  getRef(props, context) {
+    return path =>
+      this.getFirebaseApp(props, context)
+        .database()
+        .ref(path)
+  }
+
+  getFirebaseApp(props = this.props, context = this.context) {
+    return props.firebaseApp || context.firebaseApp || firebase.app()
+  }
+
+  getFirebaseProps() {
+    if (typeof this.props.query === 'string') {
+      return this.state.subscriptionsState[this.props.query]
+    }
+
+    return {
+      ...getActions(this.props, this.getRef(), this.getFirebaseApp()),
+      ...this.state.subscriptionsState,
+    }
+  }
+
+  subscribe(subscriptions, ref = this.getRef()) {
+    if (Object.keys(subscriptions).length < 1) {
+      return
+    }
+
+    const queries = mapSubscriptionsToQueries(subscriptions)
+    const nextListeners = mapValues(queries, ({ path, ...query }, key) => {
+      const containsOrderBy = Object.keys(query).some(queryKey => queryKey.startsWith('orderBy'))
+      const subscriptionRef = createQueryRef(ref(path), query)
+      const update = snapshot => {
+        if (this.mounted) {
+          const value = containsOrderBy ? mapSnapshotToValue(snapshot) : snapshot.val()
+
+          this.setState(prevState => ({
+            connected: true,
```

Comment:
If you query to multiple subscriptions like ```{ toods: 'todos', user: `users/${userId}`}``` can we be sure they are "fetched" at the same time on first connect?